### DISCUSSION TITLE
Neural network to control AI trains/AI dispatching

### DISCUSSION THREAD

**Discussion by JezSonic** [MaSzyna-Reloaded/MaSzyna-API-wrapper#274] [Ideas]
### Overview & Architecture Philosophy

This idea aims to implement an autonomous, multi-agent railway traffic management and train operation system. Instead of relying on runtime Reinforcement Learning (RL) or stochastic trial-and-error policies, the system should use **deterministic pre-trained models** paired with a **rule-based validation layer** to guarantee full compliance with safety-critical standards, specifically the **PKP PLK S.A. Ir instructions (Instrukcja o prowadzeniu ruchu pociągów)**.

Each agent functions as an independent neural entity that evaluates state spaces and predicts actions, which are then passed through an outcome-analysis filter before execution to ensure operational safety and regulatory alignment.

Agents must only use in-game commands or future dedicated API layer for them to ensure compatibility and more natural, player-like, behaviour

---
### Core Agent Hierarchy

* **Train Driver Agent (Individual Neural Network per Train):**
  * **Startup & Procedures:** Operates via a hybrid control strategy where engine startup, circuit breaker engagement, pantograph control, and brake tests execute deterministically via hard-coded simulator commands.
  * **Speed Profile & Traction Control:** The pre-trained model handles dynamic acceleration, braking curves, speed limits, signal compliance, and sudden track events (e.g., unexpected obstacles or emergency brake applications).
* **Train Dispatcher Agent (Neural Network per Station / Dispatch Area):**
  * Manages routing, switch layout configurations, interlocking systems, track occupancy, and scheduling priorities.
  * Dynamically communicates with individual Train Driver Agents to issue directives, clearances, and emergency orders.

---
### Decision-Making Pipeline: Predictive Evaluation & Ir Rule Layer

To prevent hallucinations, unsafe edge-case actions, or violations of railway law, neural networks do not directly execute low-level controls without a verification phase:

1. **State Observation:** Agent receives environmental and telemetry vectors (aspects, speed restrictions, block occupancy, communication packets).
2. **Pre-Trained Inference:** The neural model predicts the optimal action or trajectory.
3. **Outcome Analysis & Validation Layer:**
   * Simulates the immediate outcome of the proposed action.
   * Validates the candidate action against **Ir-1** (Train Movement Operations), **Ie-1** (Signaling), and related PKP PLK operational rules (e.g., substitute signals *Sz*, written orders, wrong-track running).
4. **Execution / Fallback:** If the predicted outcome is compliant, the command is committed. If a safety boundary is violated, deterministic safety fallbacks (fail-safe braking or route hold) are triggered with notification to players

---
### Required Integration API

Conducting end-to-end, validated testing requires a robust simulator API interface covering:

* **Interlocking & Infrastructure Control (SRK):**
  * Reading and overriding signal aspects and warning boards.
  * Turnout/switch monitoring and control (including detection of trailing points/interlocking state).
  * Level crossing barriers and signaling states.
* **Train Telemetry & Sensor Data:**
  * Real-time readouts of speed, brake pipe/cylinder pressures, traction currents, and vehicle parameters.
  * Spatial lookahead vectors: distances to upcoming signals, indicators, speed zones, and preceding traffic.
* **Agent-to-Agent Communication Bus:**
  * Dedicated low-latency protocol enabling structured message exchange between Dispatcher Agents and Train Driver Agents (radio orders, clearance requests, emergency stops).

---
### Next Steps & Feedback

* Implementation of the safety validator pipeline wrapping pre-trained neural outputs.
* Benchmarking data serialization and IPC transport (e.g., Shared Memory vs. gRPC) for multi-agent simulation workloads.

Contributions, architectural feedback, and technical suggestions are welcome!

**Links to instructions:**
- Ir: https://www.plk-sa.pl/klienci-i-kontrahenci/akty-prawne-i-przepisy/instrukcje-pkp-polskich-linii-kolejowych-sa/ruch-i-przewozy-kolejowe
- Ie: https://www.plk-sa.pl/klienci-i-kontrahenci/akty-prawne-i-przepisy/instrukcje-pkp-polskich-linii-kolejowych-sa/automatyka-i-telekomunikacja

**Comment by marcinn** [comment]
Can you add links/attach the documents mentioned in the post?

**Reply by JezSonic** [reply]
Sure, https://www.plk-sa.pl/klienci-i-kontrahenci/akty-prawne-i-przepisy/instrukcje-pkp-polskich-linii-kolejowych-sa/ruch-i-przewozy-kolejowe and https://www.plk-sa.pl/klienci-i-kontrahenci/akty-prawne-i-przepisy/instrukcje-pkp-polskich-linii-kolejowych-sa/automatyka-i-telekomunikacja

I've added them to the post as well

**Comment by JezSonic** [comment]
Worth mentioning is the control braking procedure, where drivers in real life, after getting on the train, slow it down normally after leaving the station to 60-40 km/h to see how it brakes. Because every train breaks different and it actually depends on the train wagon layout (they will have different brakes).

Wagons can have: disc brakes or brake pads. Additionally, with magnetic brake shoe. Different layout of wagons in the trainset therefore will change the way the whole train brakes, due to the force of inertia.

So implementing this may help the neural network with braking, as it does help train drivers in real life

**Comment by JezSonic** [comment]
### Architecture
Here's a more concrete breakdown of how that architecture should be structured — not a specific codebase, just the shape any implementation of this idea needs to take to actually be trustworthy.

The core principle: **three layers, strictly separated**, so a mistake in one can never silently propagate into the next.

```
   Neural inference  (per-train driver policy / per-scenario dispatcher policy)
                          |
                          v
   Deterministic rule layer  (regulation validator, capability/precondition checks)
                          |
                          v
   Command execution surface  (the only layer allowed to touch the simulator)
```

Nothing in the top layer is ever allowed to reach the bottom layer directly. Every action a network proposes passes through the middle layer first.

---
### Layer 1 — A vehicle-agnostic action/observation space

A fixed-shape neural network needs a fixed-length input/output vector, but not every train exposes the same subsystems (a diesel unit has fuel/oil pumps, an EMU doesn't; not every vehicle has a pneumatic spring brake). Hard-coding per-vehicle-type inference defeats the entire point of training one reusable policy.

The clean solution is a **capability descriptor** — each controllable subsystem declares itself independently of any specific model, and the model's schema is just an ordered list of capability ids it expects:

```jsonc
// A capability catalogue entry, boolean-kind:
{ "id": "spring_brake_active", "kind": "boolean",
  "state_key": "spring_brake_engaged", "command": "spring_brake_set" }

// continuous-kind (already-normalized 0..1 value, no separate min/max step):
{ "id": "brake_service_level", "kind": "continuous",
  "state_key": "brake_position_normalized", "command": "brake_level_set" }

// relative_step-kind, for controls that only support Increase/Decrease commands
// rather than an absolute "set position" (e.g. a traditional master controller):
{ "id": "main_controller", "kind": "relative_step",
  "state_key": "controller_position", "config_max_key": "controller_position_max",
  "increase_command": "controller_increase", "decrease_command": "controller_decrease",
  "max_step_per_tick": 3.0 }
```

A model's own schema is then just:

```jsonc
{ "capabilities": ["battery", "main_controller", "brake_service_level", "direction"],
  "layers": [ {"out_features": 16, "activation": "tanh"},
              {"out_features": 16, "activation": "tanh"},
              {"out_features": 8,  "activation": "tanh"} ] }
```

Building the observation vector means walking this capability list and, for any vehicle missing a given subsystem, zero-filling that slot with an explicit "not present" flag rather than erroring — and the same presence check happens again, live, when applying the model's output back to the vehicle, never trusting whatever composition existed at training time. That's the whole mechanism that lets one trained policy drive vehicles with different equipment without retraining per vehicle type.

Startup/shutdown sequencing (power on, build air pressure, engage main switch, release brakes) doesn't belong in the neural layer at all — it's a deterministic checklist that either finds a given capability present and executes it, or skips it and moves on. Save the network for genuinely ambiguous judgment calls, not procedures that already have one correct order.

---
### Layer 2 — Inference must never block the simulation

Whatever engine or framework hosts the simulation almost certainly isn't safe to poke from a background thread mid-tick. The split should happen exactly at the boundary where numeric computation begins, not anywhere else:

1. **Simulation thread**: resolve which model to run, build the observation vector, hand both off to a queue. Return immediately — never wait on the result here.
2. **Worker thread**: pop the ticket, run pure numeric inference. No calls back into simulation state from this thread, ever.
3. **Simulation thread again**: decode the result, re-check capability presence, apply the action, notify whatever is listening for the outcome.

```jsonc
// request
{ "op": "driver_action", "train_id": "train1", "model": "driver_policy" }

// response, delivered asynchronously — never as a blocking return value
{ "request_id": 42, "train_id": "train1", "ok": true,
  "capabilities": ["battery", "main_controller", "..."],
  "action_values": [1.0, 0.42, "..."] }
```

Treat the request/response pair as a message, not a function call. Any interface that makes this synchronous will eventually stall the simulation on a slow inference call.

---
### Layer 3 — The dispatcher: a scenario in, decision out

A dispatcher's input and output are not the same shape the way a driver's observation/action pair often can be — a scenario and a movement-authority decision are genuinely different structures, so the schema should be asymmetric on purpose:

```jsonc
{ "input_fields":  ["block_occupied", "block_reserved_for_other_train", "route_conflict",
                     "signal_aspect", "requested_max_speed_kmh", "written_order_issued",
                     "level_crossing_secured", "is_maneuvering_movement",
                     "matches_scheduled_timetable", "track_closed_for_maintenance", "..."],
  "output_fields": ["granted", "authorized_speed_limit_kmh"],
  "layers": [ {"out_features": 8, "activation": "tanh"},
              {"out_features": 2, "activation": "linear"} ] }
```

Every field in the scenario vocabulary should trace back to a specific instruction, not a guess — block occupancy and route conflicts, written orders that permit movement past a stop-indicating aspect, maneuvering-movement rules, level-crossing security, dangerous-goods handling, track closures, degraded-weather operating rules, and so on. A scenario is nothing more than a flat structured record over that vocabulary:

```jsonc
{ "block_occupied": false, "signal_aspect": 5, "requested_max_speed_kmh": 100,
  "route_conflict": false, "level_crossing_secured": true,
  "matches_scheduled_timetable": true }
```

---
### The layer that actually guarantees compliance

This is the part where the mechanism matters more than any specific rule encoded in it: **a network's decision must never be trusted on its own.**

Every dispatcher request should run through two independent paths in parallel:

- the trained network, producing a proposed decision;
- a deterministic, rule-based reference — a plain, auditable set of ordered checks against the same scenario, with no learned component at all.

```jsonc
// what a rule-based reference decision looks like — no network involved,
// just an ordered evaluation against the regulation text:
// 1. is the destination block closed for maintenance?         -> deny
// 2. is there an active emergency response in the area?       -> deny
// 3. is the block occupied, reserved, or route-conflicted?     -> deny
// 4. is this a maneuvering movement without a written order?   -> deny
// 5. does the signal aspect permit movement at all?             -> deny if not
// 6. otherwise: grant, with a speed cap derived from the aspect
```

The two outputs get compared, and the comparison is asymmetric on purpose — it only ever tightens, never loosens:

```jsonc
{ "network_decision":    { "granted": true,  "authorized_speed_limit_kmh": 100 },
  "regulation_decision": { "granted": true,  "authorized_speed_limit_kmh": 60  },
  "final_decision":      { "granted": true,  "authorized_speed_limit_kmh": 60  },
  "regulation_match": false,
  "regulation_overridden": true,
  "dangerous": false }
```

If the network is less restrictive than the rule-based reference — it grants what the reference denies, or grants at a higher speed than the reference's cap — the final decision gets silently replaced with the reference's, but nothing about that replacement is silent to a human: the raw network output, the reference's own answer, and whether an override happened should all be preserved in the result. A network decision that's already at least as conservative as the reference is left untouched — the rule is "never less safe than the reference," not "always ignore the network."

The one metric that should gate whether a model is allowed to run at all is the count of scenarios where the network authorized movement that the reference says must be denied. Any non-zero value on a held-out evaluation set is a blocking failure, independent of overall accuracy — a model that's 95% accurate but wrong specifically in the "grants when it shouldn't" direction is categorically more dangerous than one that's 95% accurate and only ever errs by being overly conservative.

---
### Training loop, end to end

```jsonc
// a labeled training case: exactly the shape a held-out evaluation set also uses
{ "input":    { "block_occupied": false, "signal_aspect": 5,
                "requested_max_speed_kmh": 100, "route_conflict": false },
  "expected": { "granted": true, "authorized_speed_limit_kmh": 100 } }
```

The pipeline this implies:

1. Extract plain text from the regulation documents — purely mechanical, fully automatable.
2. A human who actually understands the regulations writes labeled scenario/expected-decision pairs from that text. This step cannot be automated away — deriving a regulation-correct label from raw legal text is a judgment call, not a transformation.
3. Split labeled cases into a training set and a held-out set that is never trained on.
4. Train a small model — these are structured, low-dimensional inputs, not language, so a small network is almost always the right starting point; grow capacity only if training-set fit is genuinely underfitting.
5. Evaluate against the held-out set, not the training set. The metric that matters is the dangerous-false-grant count described above, not raw accuracy.
6. If it's not good enough, go back to step 2 and add more labeled examples near whatever the held-out evaluation got wrong — a small model like this is almost always data-limited before it's architecture-limited.

---
### What a real implementation still needs before any of this is live

None of the above works without two things that are infrastructure, not AI:

1. **A runtime track-topology graph** — blocks/sections as nodes, physical connections as edges, switch positions as routing choices. Without this, "occupied," "reserved," and "route conflict" are just fields typed in by hand rather than facts derived from an actual interlocking model.
2. **A live signal-state source** — real aspect/occupancy/reservation data feeding the scenario vocabulary above, rather than synthetic or hand-authored values.

Everything else — the capability abstraction, the async inference split, the asymmetric dispatcher schema, and above all the deterministic safety gate — should be built so that plugging a real topology/signal source in later only changes *how a scenario gets constructed*, never the shape of the schema, the service, or the validation layer around it. Design the contract first; let the data source be swappable.
